proposals/0621-linear-constraints.rst

```diff
@@ -0,0 +1,1100 @@
+some nice guarantees which ensure memory safety. For example, there is
+no way we can go on to erroneously read from the array after
+de-allocating it. Instead, when ``arr0`` is used as an argument to
+``read``, ``read`` consumes it and returns the array under a new name
+``arr1``. This process then repeats before finally ``free``
+de-allocates the array.
+
+While this code ensures that we use the ``MArray`` in a memory-safe
+way, this re-naming process introduces boilerplate and becomes
+cumbersome to both write and read. Compare to the doing the same with
+the ``ST`` monad instead:
+
+::
+
+   read2AndDiscardST :: MArray s a -> ST s (a, a)
+   read2andDiscardST arr = do
+     x <- read arr 0
+     y <- read arr 1
+     return (x, y)
+
+There is much less noise. To be honest, this is all a little silly:
+the name management is completely mechanical; it should be handled
+automatically by the compiler. This is what this proposal lets us do:
+get pretty close to the latter style, without having to introduce a
+monad.
+
+A more complex example involving freezing nested mutable structure is
+elaborated in `this blog post <blog_freeze_>`_.
+
+Creating linear values
+^^^^^^^^^^^^^^^^^^^^^^
+
+A well documented difficulty, when writing API for mutable data as
+above, is to guarantee that, say, an array is unique, it isn't
+sufficient that ``read`` and ``write`` be linear functions. If I
+create an array with
+
+::
+
+   new :: Int -> MArray -- or Int %1 -> MArray
+
+Then ``new 57`` can be shared arbitrarily. This is a phenomenon known
+as “promotion”: expressions without linear free variables are
+unrestricted. The typical solution is for ``new`` to take a
+continuation as an argument
+
+::
+
+   new :: Int -> (MArray %1 -> Ur a) %1 -> Ur a
+
+This forces the array to be single-threaded (thanks to the ``Ur a``
+return type, the ``MArray`` cannot escape the continuation's scope),
+which we can use to guarantee uniqueness.
+
+This is a little clumsy to program with. But more importantly, these
+continuations aren't very composable as argued in `this blog post
+<blog_scopes_>`_. See also the long discussion at
+`tweag/linear-base#130
+<https://github.com/tweag/linear-base/issues/130>`_. This proposal
+will let us define ``new`` in direct style.
+
+Proposed Change Specification
+-----------------------------
+
+Main Syntax Changes
+^^^^^^^^^^^^^^^^^^^
+
+Currently, type class constraints in GHC do not support multiplicty
+annotations.  GHC currently defines the syntax for type signatures as:
+
+::
+
+   ctype   ::= context '=>' ctype | type | ...
+
+Essentially, type signatures can consist of (among other things which
+we ignore here) zero or more qualified type arrows ``=>`` followed by
+a type. Here ``context`` is a list of class constraints.
+
+When ``-XLinearTypes`` is enabled, the following new syntax is
+enabled:
+
+::
+
+   ctype ::= context '%' 1 '=>' ctype | context '=>' ctype | type | ...
```

Comment:
But what if I want to have one linear and one ordinary constraint with multiplicity "`Many`"? What if I write:

```haskell
foo :: (C, Integral a) %1 => a %1 -> a %1 -> Integer`
foo a1 a2 = toInteger a1 + toInteger a2 + toInteger useC
```
I've used each `a` argument once, `C` once, but `Integral a` twice. But I think this would be forbidden by your proposal, and the only way to support it is to change the context to `(C, Integral a, Integral a) %1`. You say later that a combination is possible, but I find it hard to understand how. Maybe `C %1 => Integral a => a %1 -> a %1 -> Integer` is allowed by your syntax? That looks strange to me.

It is strange, in my opinion, that the `%1` is attached to the whole constraint context and not just a single type class. E.g. I'd expect `(C %1, C %1) => Int` instead of `(C, C) %1 => Int`. If this was the proposed syntax, then I could also combine linear and normal constraint with `(C %1, OtherConstraint) => ...`, and use `OtherConstraint` as often as I want.

Maybe I'm also misunderstanding something here, but that might then be a sign that this case could be clearer ;)

Reply:
Under this proposal, you'd write

```haskell
foo :: C %1 => Integral a => a %1 -> a %1 -> Integer
```

There is no theoretical obstacle here. I proposed the most minimalistic syntax change, largely because I'm scared of GHC's parser. There's also no support for unnamed tuples with fields of varying multiplicity in Core, and I believe (though I haven't checked) that these tuples of constraints are desugared as tuples. Anything that doesn't change Core makes my life much easier.

See [_Unrestricted modifier for constraints_](https://github.com/tweag/ghc-proposals/blob/linear-constraints/proposals/0621-linear-constraints.rst#74unrestricted-modifier-for-constraints) in the _Alternatives_ section.